src/dag.rs

```diff
@@ -995,3 +1027,124 @@ impl<D: DagLike, S: SharingTracker<D>> Iterator for PreOrderIter<D, S> {
         None
     }
 }
+
+/// Iterates over a DAG in _pre order_, yielding extra state changes.
+///
+/// Unlike the regular pre-order iterator, this one may yield each node
+/// multiple times: once in pre-order and then again after each of its
+/// children has been yielded. The different cases can be distinguished
+/// by inspecting the [`PreOrderIterItem::n_children_yielded`] and
+/// [`PreOrderIterItem::is_complete`] fields of the yielded item.
+///
+/// If you use this iterator with a non-trivial sharing tracker, then any
+/// items which have been initially yielded before will not be initially
+/// yielded again.
+///
+/// If node's *children* have been initially yielded before, they will be
+/// skipped, but the node itself will still be re-yielded as many times
+/// as it has children.
+#[derive(Clone, Debug)]
+pub struct VerbosePreOrderIter<D: DagLike + Clone, S: SharingTracker<D>> {
+    /// A stack of elements to be yielded. As items are yielded, their right
+    /// children are put onto the stack followed by their left, so that the
+    /// appropriate one will be yielded on the next iteration.
+    stack: Vec<PreOrderIterItem<D>>,
+    /// The index of the next item to be yielded.
+    ///
+    /// Note that unlike the [`PostOrderIter`], this value is not monotonic
+    /// and not equivalent to just using `enumerate` on the iterator, because
+    /// elements may be yielded multiple times.
+    index: usize,
+    /// Data which tracks which nodes have been yielded already and therefore
+    /// should be skipped.
+    tracker: S,
+}
+
+impl<D: DagLike + Clone, S: SharingTracker<D>> Iterator for VerbosePreOrderIter<D, S> {
+    type Item = PreOrderIterItem<D>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // This algorithm is _significantly_ simpler than the post-order one,
+        // mainly because we don't care about indices.
```

Comment:
87aadf914254570e099522f452942ce8e1d82f0a: copy-paste error: the verbose pre-order iterator does care about indices

Reply:
Should probably say "child indices" in both cases.